src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/RazorDynamicFileInfoProvider.cs

```diff
@@ -46,19 +53,41 @@ private class RemoveDynamicFileParams
 #pragma warning restore CS0067
 
     private readonly Lazy<RazorWorkspaceListenerInitializer> _razorWorkspaceListenerInitializer;
+    private readonly Lazy<LanguageServerWorkspaceFactory> _workspaceFactory;
 
     [ImportingConstructor]
     [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-    public RazorDynamicFileInfoProvider(Lazy<RazorWorkspaceListenerInitializer> razorWorkspaceListenerInitializer)
+    public RazorDynamicFileInfoProvider(Lazy<RazorWorkspaceListenerInitializer> razorWorkspaceListenerInitializer, Lazy<LanguageServerWorkspaceFactory> workspaceFactory)
     {
         _razorWorkspaceListenerInitializer = razorWorkspaceListenerInitializer;
+        _workspaceFactory = workspaceFactory;
     }
 
     public async Task<DynamicFileInfo?> GetDynamicFileInfoAsync(ProjectId projectId, string? projectFilePath, string filePath, CancellationToken cancellationToken)
     {
+        var project = _workspaceFactory.Value.Workspace.CurrentSolution.GetRequiredProject(projectId);
```

Comment:
Do we ensure this project can't go away during this call? I guess we're calling under the project's lock, but this seems a bit worrisome. What happens if we just don't do this?

Reply:
I'm happy to just bail out of here if the project is null.

Reply:
I had to think a bit about whether this was even safe or whether due to batching you could end up in bad places. I think it's fine, but I guess I wasn't sure what problem this change was trying to solve in the first place.

Reply:
Also because practically we only will ever have C# projects in our workspace, so... :smile:

Reply:
> practically we only will ever have C# projects in our workspace

The language check is not super important, consider this future proofing so we don't reitroduce https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1878509

I figured since I needed the project anyway, to get the intermediate output path, I may as well put the check in.